add lock in rh/genimage to avoid multiple genimage processes on same rootimg directory;add guard code before umount/remove mount points under rootimg directory

immarvin opened the pull request:
fix https://github.com/xcat2/xcat-core/issues/1936
fix https://github.com/xcat2/xcat-core/issues/1767
1. add lock in rh/genimage to avoid multiple genimage processes on same rootimg directory;
2. add guard code for umount/remove mount points under rootimg directory

daniceexi commented:
@immarvin Good enhancement.

All of xcat commands do not support ``lock`` that avoid to be run in parallel. Could we make this ``lock`` mechanism code to be a general subroutines in the Utils.pm so that other cmd also can use it.

neo954 commented on xCAT-server/lib/xcat/plugins/rmimage.pm:
```diff
@@ -185,12 +186,20 @@ sub process_request {
+    #unmount all the mount points under rootimg directory 
+    #to avoid removing the directory/files on management node by mistake  
+    $realimagedir=realpath($imagedir);
+    my @mntptlist=qx(cat /proc/mounts |awk -F' ' '{print \$2}'|grep "$realimagedir" 2>/dev/null);
```
Since we are writing perl code here. I suggest do not use external command here. This makes the code more complex and less efficiency. Please use perl to do things here.

That is something like the following ...

    open FILE, "<", "/proc/mounts" or die;
    while (<FILE>) {
        my @arr = split / /;
        print $arr[1] if (substr($arr[1], 0, length($realimagedir)) eq $realimagedir);
    }
    close FILE;

neo954 replied:
Consider the `$realimgedir` is `/path/to/somewhere`. But in `/proc/mounts` there is a line `/some/other/dir/path/to/somewhere/foo/bar`. In this case, use `grep` is not really good enough. That is why I use `substr()` and string comparison in the example code above.

neo954 replied:
And please also consider if there are `/path/to/somewhere` and `/path/to/somewhere-else`, then, while you matching `/path/to/somewhere`, try not `umount`ing subdirectories under `/path/to/somewhere-else`

neo954 commented on xCAT-server/share/xcat/netboot/rh/genimage:
```diff
@@ -66,6 +66,38 @@ my $tempfile;
+    open($lockinfo,">", $lockinfofn) or return 1;
+    print $lockinfo  "$$";
+    close($lockinfo);
+    if( -l $lockfile){
```
Should check if `$lockfile` is `undef` first. Or maybe, `$lockfile` should be passed as an argument of this sub-function.

neo954 commented on xCAT-server/share/xcat/netboot/rh/genimage:
```diff
@@ -66,6 +66,38 @@ my $tempfile;
+      open($lockinfo,"<",$lockfile ) or return 1;
+      $exps=<$lockinfo>;
+      close($lockinfo);
+      system qq(ps --pid $exps --format command h|grep -i genimage) or return 2;
```
Please do not use external command, but use `kill SIGZERO $exps;` to check if the process exist.

neo954 replied:
And if the process `$exps` does not exist here, the program should on, not `return 2` here.

neo954 commented on xCAT-server/share/xcat/netboot/rh/genimage:
```diff
@@ -66,6 +66,38 @@ my $tempfile;
+      $exps=<$lockinfo>;
+      close($lockinfo);
+      system qq(ps --pid $exps --format command h|grep -i genimage) or return 2;
+      unlink($lockfile);
```
Why delete the file `$lockfile` here. This is not right.

neo954 commented on xCAT-server/share/xcat/netboot/rh/genimage:
```diff
@@ -66,6 +66,38 @@ my $tempfile;
+    print $lockinfo  "$$";
+    close($lockinfo);
+    if( -l $lockfile){
+      open($lockinfo,"<",$lockfile ) or return 1;
```
If the file `$lockfile` does not exist, the program should move on, not `return  1` here.

neo954 commented on xCAT-server/share/xcat/netboot/rh/genimage:
```diff
@@ -66,6 +66,38 @@ my $tempfile;
+        $ret=link($lockinfofn, $lockfile);
+    } 
+    $ret=($ret==0) ? 1:0;
+    return $ret;
```
In this case, `return not $ret;` will do the work.

neo954 commented on xCAT-server/share/xcat/netboot/rh/genimage:
```diff
@@ -112,17 +144,26 @@ sub mount_chroot {
+    
+    #unmount all the mount points under rootimg directory
+    #to avoid removing the direcoty/files on managemment node by mistake
+    my @mntptlist=qx(cat /proc/mounts |awk -F' ' '{print \$2}'|grep "$rootimage_dir" 2>/dev/null);
```
Same as the comment above, please use perl codes to do the work.

gurevichmark commented:
Since this pull request makes changes to the rmimage.pm, will this be the good place to also fix the problem that `rmimage` command does not remove the `rootimg.cpio.gz` file ? I believe during one of our conference call it was agreed that this command should do that.

whowutwut commented on xCAT-server/lib/xcat/plugins/rmimage.pm:
```diff
@@ -185,12 +186,20 @@ sub process_request {
 
     #some rpms like atftp mount the rootimg/proc to /proc, we need to make sure rootimg/proc is free of junk
-    `umount -l $imagedir/rootimg/proc 2>&1 1>/dev/null`;
+    #`umount -l $imagedir/rootimg/proc 2>&1 1>/dev/null`;
```
Should delete these lines instead of comment out.

immarvin commented:
hi @whowutwut @gurevichmark @daniceexi @neo954 , all the comments have been processed

neo954 commented on xCAT-server/share/xcat/netboot/rh/genimage:
```diff
@@ -348,6 +366,18 @@ unless ($onlyinitrd) {
+    my $NON_BLOCK=1;
+    $lock = xCAT::Utils->acquire_lock("$mylockfile", $NON_BLOCK);
+    unless ($lock){
+        my $pid=qx(lsof -Fp /var/run/lock/xcat/$mylockfile 2>/dev/null|tr p " ");
```
Package `lsof` is not installed by default. If you insist use `lsof` here, you have to put package `lsof` into the dependency list of package `xCAT-server`.

neo954 commented on xCAT-server/share/xcat/netboot/rh/genimage:
```diff
@@ -112,17 +116,30 @@ sub mount_chroot {
-        #system("umount -l $rootimage_dir/dev");
-        system("rm -rf $rootimage_dir/dev/*");
+        #only remove the /dev in rootimg directory if it is not a mount point
+        system("findmnt $rootimage_dir/dev/ >/dev/null 2>&1 || rm -rf $rootimage_dir/dev/*");
```
Even if udev is used, remove everything under the `/dev` directory may cause problem. At lease `/dev/null`, `/dev/zero` should be there.

immarvin replied:
the amount will occur on genimage exit or interrupted, no following operation will be performed on rootimg,  I think it's ok to remove the /dev/null and /dev/zero at this time

neo954 commented on xCAT-server/lib/xcat/plugins/packimage.pm:
```diff
@@ -199,6 +202,21 @@ sub process_request {
+    my $mylockfile=Cwd::realpath($rootimg_dir);
+    my $mymd5=md5_hex($mylockfile);    
+    $mylockfile=~s/\//./g;
+    my $mylockfile=$mylockfile.".".$mymd5;
```
I still prefer let `$mylockfile = "../../.." . $destdir . "/.genimage.lock";`

immarvin replied:
we  have talked about this offline, the `/var/run/lock/xcat/` is the conventional place of locks in xCAT

neo954 commented on xCAT-server/lib/xcat/plugins/rmimage.pm:
```diff
@@ -175,6 +178,22 @@ sub process_request {
+        close($pidfd);
+        $callback->({ error => ["failed to acquire lock, seems there is another genimage/packimage/rmimage process $pid running on root image dir \"$rootimg_dir\""], errorcode => [1] });
+        return 1;
+    }
```
There are identical codes in `packimg.pm`, `rmimage.pm` and `genimage`. I think there is no point to write the same code three times.

I suggest made these codes a function, and reuse the same code in three different places.

immarvin replied:
good idea, but there is a problem that the lock will be released outside the namespace which  "$lock" in,  in the following code slice
````
$lock = xCAT::Utils->acquire_lock("$mylockfile", $NON_BLOCK);
````

with this problem, we'd better not to abstract this code slice as a subroutine.
I leave this as a issue https://github.com/xcat2/xcat-core/issues/2078

neo954 commented on xCAT-server/share/xcat/netboot/rh/genimage:
```diff
@@ -348,6 +366,21 @@ unless ($onlyinitrd) {
+        print "genimage: failed to acquire lock, seems there is another genimage/packimage/rmimage process $pid running on root image dir \"$destdir\" ";
+        exit 1;
+    }
+
```
It is worth to port the "lock" feature to Debian/Ubuntu as well.

immarvin replied:
will open another issue on this, do not spare time on the unit test on sles and ubuntu this PR

immarvin replied:
traced wth the ticket https://github.com/xcat2/xcat-core/issues/2077

neo954 commented:
I believe this patch is good enough to merge. Great works, @immarvin.

I will merge this one.

immarvin commented:
hi @daniceexi , there is already some lock subroutines(acquire_lock,....) in Utils.pm,  I leveraged the subroutine in this PR.

These are several issues in this subroutine, I fixed them in this  PR.

I added the lock for packimage, rummage and rh/genimage in this PR. the genimage on sles and ubuntu, imgimport, imgexport and liteimg  are left in ticket https://github.com/xcat2/xcat-core/issues/2077.

hi @gurevichmark , I have also fixed the issue that rmimage can not find rootimg.gz you mentioned in this PR

immarvin commented:
@neo954 raised a lot of valuable comments on this PR to help refine the code, and provided a lot of support on technical details and suggestion on debug, thanks a lot

daniceexi commented:
@immarvin Thanks 👍

daniceexi commented:
@neo954 Good work. Thanks 👍
